plugins/sla_check.py

```diff
@@ -45,82 +42,84 @@
+  template_folder='templates',
+  static_folder='static',
+  static_url_path='/static/sla_check')
+
 
 class SLACheckView(BaseView):
-  # Diagnostic endpoint which lists out the discovered SLA's and
-  # what time they are due at today.
   @expose('/')
   def index(self):
     logging.info("SLA Check index() called")
     slas = Variable.get(SLA_CONFIG_VARIABLE, default_var=dict(), deserialize_json=True)
-    print slas
     return self.render("sla_check/index.html", slas=slas)
 
-  # Wraps the violations for DataDog's format, it will return 200 if
-  # there are no violations and 500 and a comma-separated string of
-  # dag_id's that are currently in violation.
   @expose('/slacheck')
   def sla_check(self):
-    logging.info("SLA Check() called")
+    logging.info("SLA Check called")
+
+    # load sla configuration
+    slas = Variable.get(SLA_CONFIG_VARIABLE, default_var=list(), deserialize_json=True)
+    logging.info("Loaded SLA configuration: {}".format(slas))
+
+    dagbag = DagBag()
+
+    logging.info("DAG keys: {}".format(dagbag.dags.keys()))
+
+    # get all dags with an SLA defined in the slas variable
+    dags_slas = [(dagbag.dags[s['dag_id']], s['duration']) for s in slas if s['dag_id'] in dagbag.dags.keys()]
+    logging.info("DAGs with an sla defined {}".format(dags_slas))
 
     violations = []
 
-    # Load SLA's and DAG configurations
-    slas = Variable.get(SLA_CONFIG_VARIABLE, default_var=dict(), deserialize_json=True)
-    logging.info("Loaded SLA configuration for {0}".format(slas))
+    # for each of these dags with SLAs, find the last run which should have happened
+    dags_last_runs = []
+    for dag, sla_max_duration in dags_slas:
+      if dag.schedule_interval and not dag.is_paused:
+        dttm_now = datetime.now()
 
-    dagbag = DagBag()
+        latest_execution_run_date, following_schedule_date = _get_run_dates(dag, timedelta(minutes=sla_max_duration))
+
+        # check if execution date + schedule_interval + sla_duration is after the current time
+        sla_due_date = following_schedule_date + timedelta(minutes=sla_max_duration)
+
+        logging.info("execution_date={0} for dag_id={1} with an SLA={2} has due_date={3}"
+                     .format(latest_execution_run_date, dag.dag_id, sla_max_duration, sla_due_date))
+
+        if dttm_now > sla_due_date:
+          last_dag_run = settings.Session.query(DagRun).filter(
+            DagRun.dag_id == dag.dag_id,
+            DagRun.execution_date == latest_execution_run_date).first()
+
+          # report if not successful
+          if not last_dag_run or last_dag_run.state != State.SUCCESS:
+            logging.info("Found that the SLA for {0} is in violation of configured SLA".format(dag.dag_id))
+            violations.append(
+              {"dag_id": dag.dag_id,
+               "due_date": sla_due_date.isoformat(),
+               "execution_date": latest_execution_run_date.isoformat(),
+               "sla_duration": sla_max_duration
```

Comment:
would it make sense to have the state of the last run here? We might care if it failed vs if it was canceled.